### PR TITLE
Extend prpl-server with SEO magic

### PR DESCRIPTION
**Pull request by WillMoggridge**
This extends the prpl-server used in production. It adds mittleware to set headers and HTML head.

The key features:
- Send correct 404 headers on non existant `/tutorials/*` and `/api/*` URLs.
- Set metadata in the output, looking up correct data from codelabs API.
- Add a `sitemap.xml` to link to every tutorial, and a robots.txt to link to sitemap

## QA
QA'ing this will require not using the ./run script.

- `yarn run clean`
- `yarn install`
- `yarn run build-all`
- `PORT=8081 yarn run start-server`

I have been testing metadata output with:
`http http://0.0.0.0:8081/tutorial/install-openstack-with-conjure-up\#0 -p hb | rg -e 'HTTP.*$' -e '<title>[^<]*</title>' -e '<meta[^>]*>' -e '<[^<>]*canonical[^>]*>' -o`

(Using `httpie` for web request, and `rg`/`ripgrep` is a really nice grep replacement)

```
HTTP/1.1 200 OK
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1,user-scalable=yes">
<title>Install single-server OpenStack with conjure-up | Ubuntu tutorials</title>
<meta name="description" content="Learn how to deploy OpenStack on a single machine using the conjure-up deployment tool.">
<meta name="theme-color" content="#dd4814">
<meta name="google-site-verification" content="0IvYwDQk61DQg67UU7CV7LfBnPchnsRSupJp83QOCh8">
<meta name="og:title" content="Install single-server OpenStack with conjure-up | Ubuntu tutorials">
<meta name="og:type" content="article">
<meta name="og:url" content="https://tutorials.ubuntu.com/tutorial/install-openstack-with-conjure-up">
<meta name="author" content="Konrad Krawiec">
<meta name="article:author" content="Konrad Krawiec">
<meta name="article:published_time" content="2018-01-13T00:00:01Z">
<meta name="og:description" content="Learn how to deploy OpenStack on a single machine using the conjure-up deployment tool.">
<link rel="canonical" href="https://tutorials.ubuntu.com/tutorial/install-openstack-with-conjure-up">

```

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -206,6 +206,18 @@ app.use('/api', express.static(
   }
 ));
 
+// If a tutorial does not exist, return a 404 header
+app.use('/tutorial/:id', function(req, res, next){
+  let id = req.params.id;
```
Comment (tbille) [inline server.js]: use `const`

File: server.js [hunk 2]
```diff
@@ -145,9 +145,26 @@ if (args['cache-control']) {
   config.cacheControl = args['cache-control'];
 };
 
+// Route prpl-server errors back to Express
+config.forwardErrors = true;
+
+const hostname = 'https://tutorials.ubuntu.com';
```
Comment (tbille) [inline server.js]: this should have no impact on staging?

File: src/codelabs-page.html [hunk 3]
```diff
@@ -220,8 +220,15 @@
...
         if (metadata !== undefined) {
           this.fire('app-metadata', {
-            title: metadata.title + ' | Ubuntu tutorials',
-            description: metadata.summary,
+            'title': metadata.title + ' | Ubuntu tutorials',
+            'description': metadata.summary,
+            'author': metadata.author || '',
+            'og:description': metadata.summary || '',
+            'og:image': metadata.image || '',
+            'og:title': metadata.title || '',
```
Comment (tbille) [inline src/codelabs-page.html]: `'og:title': metadata.title + ' | Ubuntu tutorials',` ?

File: server.js (hunk 1 above)
Comment (WillMoggridge) [reply server.js]: Good call

File: src/codelabs-page.html (hunk 3 above)
Comment (WillMoggridge) [reply src/codelabs-page.html]: Ah, thanks!

**Pull request merged**